keps/sig-api-machinery/2608-crd-field-level-versioning/README.md

```diff
@@ -0,0 +1,816 @@
+would still get deprecation warnings asking them to switch to a new version (provided it is made 
+available).  
+
+Field level versioning for CRs will introduce the ability to have individual fields of a
+CR in different stages and aims to try and simplify the above mentioned process and provide
+a UX similar to built-in types.
+
+### Goals
+
+- Introduce custom feature gates that can be embedded as part of the CRD spec.
+- Define the lifecycle of these custom feature gates and specify what the behaviour
+  of the gate would be at different stages in its lifecycle.
+- Allow issuing custom deprecation warnings for deprecated fields.  
+- Extend the CRD validation chain to include checks for proper usage of these
+  custom feature gates.
+
+### Non-Goals
+
+- Provide a generic mechanism for dynamic feature gate functionality.
+
+## Proposal
+
+This KEP proposes to extend the CRD spec to include feature gate information
+for fields of Custom Resources (CRs):
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: crontabs.stable.example.com
+spec:
+  group: stable.example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                cronSpec:
+                  type: string
+                image:
+                  type: string
+                replicas:
+                  type: integer
+  scope: Namespaced
+  names:
+    ...
+  customFeatureGates:
+    component: istio # optional
+    featureGates:
+      - name: ReplicasFeatureGate
+        enabled: true  # optional
+        default: false # optional
+        preRelease: alpha
+        fieldPaths:
+          - .spec.replicas
+```
+This KEP, on a high level, proposes the following:
+- Multiple feature gates can be defined and each feature gate can then gate
+  multiple `fieldPaths` - each of which is a JSON path for a field in a Custom
+  Resource (CR) of this CRD.
+- Each feature gate defines the state that it is in, namely `preRelease`.
+  - The `preRelease` field can be in one of the following states:
+    - `alpha`
+    - `beta`
+    - `stable`
+    - `deprecated`
+      - Featuer gates that have `preRelease` as `deprecated` can optionally
+        specify a deprecation warning that the server will send out when this
+        field is used.
+- Additionally, a feature gate can also optionally specify if it is `enabled` or not
+  as well as optionally specify what the `default` state should be for the gate if
+  `enabled` is not specified.
+- Feature gates are evaluated and "applied" only on the storage versions of CRs whose
+  CRDs define feature gates, i.e. they are evaluated after conversion(s) to the storage
+  version take place. This behaviour is consistent with that of built-in types.
```

Comment:
is my memory correct that we convert CRs only after all validation has taken place? This asks for a complicated implementation if that is true. Is there a proof PR showing the complexity of this?

Reply:
Please clarify: we only drop on write or also on read?